### PR TITLE
Issue/2557 madx particle tracking

### PR DESCRIPTION
**Pull request by mkeilman**
Notes:

- added bunch reports
- streamlined ```_ptc_particles()``` to accept a dict containing the twiss params
- map aforementioned twiss dict to phase space variables in the schema
- general cleanup of names etc.

### CODE REVIEW

File: sirepo/package_data/static/json/madx-schema.json [hunk 1]
```diff
@@ -472,6 +479,16 @@
...
         },
+        "initialTwissParams": {
+            "alpha": ["Alpha", "Float", 0.0],
+            "beta": ["Beta", "Float", 1.0],
+            "dim": ["Dimension", "SpaceCoordinate", "x"],
+            "eta": ["Emittance", "Float", 1.0],
+            "gamma": ["Gamma", "Float", 1.0]
+        },
+        "particleTracking": {
+            "numParticles": ["Number of particles", "Integer", 1000, "", 1, 100000]
```
Comment (e-carlin) [inline sirepo/package_data/static/json/madx-schema.json:490]: in "simulation" we have "numberOfParticles". These should probably be the same value. Doesn't matter to me which one we use.

File: sirepo/pkcli/madx.py [hunk 2]
```diff
@@ -25,12 +25,29 @@ def run(cfg_dir):
         pkio.py_path(cfg_dir),
     )
 
+
 def run_background(cfg_dir):
     _run_madx()
 
+
 def _run_madx():
+    #data = simulation_db.read_json(template_common.INPUT_BASE_NAME)
```
Comment (e-carlin) [inline sirepo/pkcli/madx.py:34]: Can the commented out code be removed?

File: sirepo/sim_data/madx.py [hunk 3]
```diff
@@ -20,21 +20,40 @@ def fixup_old_data(cls, data):
...
+            m = dm['particleTracking'] = PKDict()
+            cls.update_model_defaults(m, 'particleTracking')
         for container in ('commands', 'elements'):
             for m in dm[container]:
                 cls.update_model_defaults(m, LatticeUtil.model_name_for_data(m))
 
+
     @classmethod
     def _compute_job_fields(cls, data, r, compute_model):
         # TODO(e-carlin): impl
```
Comment (e-carlin) [inline sirepo/sim_data/madx.py:45]: Remove my todo.

File: sirepo/template/madx.py [hunk 4]
```diff
@@ -270,10 +271,69 @@ def _extract_report_data(data, run_dir):
     )(data, run_dir)
 
 
+def _extract_report_bunchReport(data, run_dir):
+    util = LatticeUtil(data, _SCHEMA)
+    m = util.find_first_command(data, 'twiss')
+    if not m:
```
Comment (e-carlin) [inline sirepo/template/madx.py:277]: Twiss command is in default-data and can't be deleted in the GUI if it is the last one so it is required. I don't think we need to check.

File: sirepo/template/madx.py [hunk 5]
```diff
@@ -270,10 +271,69 @@ def _extract_report_data(data, run_dir):
...
+
+
+def _get_initial_twiss_params(data):
+    p = PKDict(
+        x=PKDict(),
+        y=PKDict()
+    )
+    util = LatticeUtil(data, _SCHEMA)
+    m = util.find_first_command(data, 'twiss')
+    # must an initial twiss always exist?
```
Comment (e-carlin) [inline sirepo/template/madx.py:318]: Yes, see comment above.

File: sirepo/template/madx.py [hunk 6]
```diff
@@ -364,8 +413,17 @@ def _extract_report_twissReport(data, run_dir):
     )
 
 
+def _format_field_value(state, model, field, el_type):
```
Comment (e-carlin) [inline sirepo/template/madx.py:416]: This is a duplicate function (see line 510).

File: sirepo/template/madx.py [hunk 7]
```diff
@@ -447,75 +516,53 @@ def _format_field_value(state, model, field, el_type):
...
+    )
+
     v = PKDict(
-        x=p[:,0],
-        px=p[:,1],
-        y=p[:,2],
-        py=p[:,3],
-        t=p[:,4],
-        pt=p[:,5],
+        x=p.x.pos,
```
Comment (e-carlin) [inline sirepo/template/madx.py:558]: Nice, thank you.

File: sirepo/package_data/static/json/madx-schema.json (hunk 1 above)
Comment (mkeilman) [reply sirepo/package_data/static/json/madx-schema.json:490]: I now see there is also a beam parameter ```NPART``` which is particles per bunch (madx default is 0 for some reason). Does that serve the same purpose or is it it's own thing?

Comment (e-carlin) [reply sirepo/package_data/static/json/madx-schema.json]: Hmm I hadn't noticed that. I think maybe it is its own thing. We use the numberOfParticles to generate the `ptc_start` commands. NPART seems to be doing something else internal with madx. 

You can see the docs [here](https://mad.web.cern.ch/mad/webguide/manual.html#Ch7). They describe NPART as "the particles in the bunch". I'm not really sure how NPART and the ptc_start commands are related.

File: sirepo/sim_data/madx.py [hunk 8]
```diff
@@ -20,21 +20,39 @@ def fixup_old_data(cls, data):
...
+                cls.update_model_defaults(m, 'bunchReport')
+                m.x = b_params[i][0]
+                m.y = b_params[i][1]
+        if 'initialTwissParams' not in dm:
+            for dim in ['x', 'y']:
+                m = dm['initialTwissParams'] = PKDict()
+                cls.update_model_defaults(m, 'initialTwissParams')
+                m.dim = dim
+        if 'particleTracking' not in dm:
+            m = dm['particleTracking'] = PKDict()
```
Comment (moellep) [inline sirepo/sim_data/madx.py:36]: I don't see the particleTracking model in the schema - was it replaced with something else and this line can be removed?

File: sirepo/template/madx.py [hunk 9]
```diff
@@ -393,6 +438,17 @@ def _generate_lattice(util):
 
 def _generate_parameters_file(data):
     res, v = template_common.generate_parameters_file(data)
+
+    v.report = re.sub(r'\d+$', '', data.report)
+    if v.report in _INITIAL_REPORTS:
+        # these reports do not require running madx first
+        v.initialTwissParameters = _get_initial_twiss_params(data)
+        v.numParticles = data.models.particleTracking.numParticles
```
Comment (moellep) [inline sirepo/template/madx.py:446]: another reference to the particleTracking model.

File: sirepo/sim_data/madx.py (hunk 8 above)
Comment (mkeilman) [reply sirepo/sim_data/madx.py:36]: Yes - there was already simulation.numberOfParticles so particleTracking model is no longer used.

**Pull request merged**